proposals/2023-08-21-utf8.md

```diff
@@ -0,0 +1,245 @@
+
+* Recording Rule Syntax: `sum(rate({"my.dotted.metric:errors:rate5m", region="east"}[5m]))`. Recording rule syntax is a naming convention, not a programmatically-meaningful operator syntax.
+
+* Unicode escaping: `{"\u1F4A9"}`
+
+* Compatible with possible future Native Histogram syntax (not being proposed here):
+  `{"my.metric.with.periods", region="east"}.sum`
+
+* Compatible with possible future metric name operator syntax via prefixes (not being proposed here):
+  `{~"a\.regex.*selector.*", region="east"}` or `{!"not.this.name", region="east"}`
+
+### Unusual / Complex Characters
+
+Depending on the parsing implementation, there may be UTF-8 characters that, when encountered as native values in quotes, may cause problems with the lexers and parsers. These characters could include whitespace, especially newlines, or complex combiner characters. If such characters exist, we will require that they be escaped with `\u` rather than allow them to exist literally. As development of this proposal progresses, we will discover which characters will fall into this category.
+
+### Templating syntax
+
+Templating is used in alerts, and for dashboards that want to reference metric values. Go templating already supports quoting by way of the `index` keyword. Whereas labels are usually denoted as `{{ $labels.my_label_name }}`, the quote syntax is `{{ index $labels "my.label.name" }}`. No code changes are required.
+
+## Backwards-compatibility
+
+There a few situations where we need to consider backwards compatibility:
+
+1. Old metrics provider, new backend: Old providers will not serve UTF-8 metric names, so nothing needs to be done here.
+2. New metrics provider, old backend: Old versions of Prometheus will throw an error if they see the new syntax. Therefore we must provide a way for the agent to serve previously-invalid metrics in a form that the old database can process.
+3. New backend, old query client: Errors will occur if an old client is fed UTF-8 metric names. Therefore we must provide a way for Prometheus to escape metric names such that old clients can query for and read them.
+4. Old backend, new query client: as with the first case, nothing needs to be done.
+
+### Text escaping
+
+We propose an escaping syntax for metric names that contain UTF-8 characters for systems that do not support UTF-8. This system can be used in queries and responses to provide compatibility between the systems.
+
+* Prefix the string with `U__`.
+* All non-valid characters (i.e. characters other than letters, numbers, and underscores) will be encoded as underscores surrounding the unicode value, like `_1F60A_`.
+* All pre-existing underscores will become doubled: `__`.
+* If a string should start with "U__" already, it will need to be escaped: `U___55_____`. (That's `U__` + `_55_` (for 'U') + `__` + `__`).
```

Comment:
I have realized now that this approach doesn't really solve all the edge cases.

If there is an existing metric name that starts with `U__`, for example `U___263a_`, and we upgrade the backend from old to new, an old querier will suddenly receive the escaped version of the existing metric name, i.e. `U___55_______263a__` rather than the old form. Or in other words: We suddenly get a different response after an upgrade that is supposed to be compatible.

We could rule that we only apply the escaping where needed, i.e. only to newly appearing names with formerly invalid characters. But that would lead to another problem: Now a new metric called `☺` would be returned to an old client as `U___263a_`, while the existing metric would be returned in the same way.

The question is how relevant this is in practice. Is there an existing name that starts with `U___` (i.e. triple underscore) and then happens to have a valid unicode hex number after it and then another underscore?

However, the problem needs to be acknowledged here.

Reply:
> We could rule that we only apply the escaping where needed, i.e. only to newly appearing names with formerly invalid characters. 

Yes that is the approach I would propose

> But that would lead to another problem: Now a new metric called `☺` would be returned to an old client as `U___263a_`, while the existing metric would be returned in the same way.

I would content that  `U___263a_` and `☺` are the same metric, just two different ways of representing it. I see your point about upgrading the backend -- a query for `☺` will not hit `U___263a_`, because the backend will look for the native character. Perhaps that means the on-disk representation needs a flag in the metadata file to indicate if the block is UTF-8-native.  The library reading the tsdb will understand that a query for `☺` against blocks that do not have support should look for  `U___263a_`. The same query, hitting a newer block, will have  `☺`.

This may not be common but this is an excellent point, we need to cover this situation or else Bad Things will happen.  (And asking people to recode all of their persisted storage is not feasible)

Reply:
As said, all of this is fairly irrelevant in practice, but we have to formulate clear rules (maybe not complete rules in the design doc, but we need to point to the areas where the rules need to be completed eventually).

Reply:
A remotely similar case was https://github.com/prometheus/prometheus/issues/5861 . The [test case](https://github.com/prometheus/prometheus/blob/83891135c6585fc2e9990824c9d12478acfe6b10/scrape/scrape_test.go#L1563-L1574) illustrates funny edge cases…

Reply:
We could bump the version number in meta.json: https://ganeshvernekar.com/blog/prometheus-tsdb-persistent-block-and-its-index/#1-metajson

older versions would indicate that U__ metrics should be unescaped.  newer versions would expect native UTF-8 characters

Reply:
code ref: https://github.com/prometheus/prometheus/blob/main/tsdb/block.go#L240